### PR TITLE
Feat: Add Same-ness Testing Functions and Improve Test Patterns

### PR DESCRIPTION
**Pull request by amery**
### **User description**

# Add Same-ness Testing Functions and Improve Test Patterns

## Summary

This PR introduces new testing utilities for same-ness comparison in Go
and modernises existing test patterns to use them. Same-ness testing
distinguishes between value equality (for basic types) and reference
equality (for reference types), filling a gap in the current testing
toolkit.

### New Functions Added

- **`IsSame(a, b any) bool`** - Core function for same-ness comparison.
  - Value types (numbers, strings, booleans): compares by equal values.
  - Reference types (slices, maps, pointers, channels, functions):
    compares by pointer equality.
  - Proper nil handling: typed nils of same type are considered same.
  - Interface support with recursive comparison of contained values.
  - Special handling for Go's runtime optimisation of empty slices.
  - Protection against invalid reflect.Values to prevent panics.

- **`AssertSame(t, expected, actual, name, args...)`** - Testing
  assertion for same references/values.
- **`AssertNotSame(t, expected, actual, name, args...)`** - Testing
  assertion for different references/values.
- **`AssertMustSame(t, expected, actual, name, args...)`** - Fatal
  assertion for same references/values that calls `t.FailNow()` on
  failure.
- **`AssertMustNotSame(t, expected, actual, name, args...)`** - Fatal
  assertion for different references/values that calls `t.FailNow()` on
  failure.

### Cross-Compatible Test Utilities

- **doRun Pattern Documentation** - Added comprehensive documentation for
  creating test functions that work with both `*testing.T` and `MockT`:
  - Interface type assertion patterns for Run method detection.
  - Graceful degradation for different testing interfaces.
  - Enhanced MockT.Run documentation with usage examples.
  - New README.md section with complete implementation guide.

### Test Pattern Improvements

Modernised 24 locations across 4 test files to follow documentation
standards:

- Replaced manual pointer comparisons with `AssertSame()`/`AssertNotSame()`.
- Converted `t.Fatal()` patterns to `if !Assert() { return }` per
  documentation.
- Improved error messages and consistency.

### Implementation Highlights

- **Complexity-compliant**: Uses helper functions to meet cognitive
  complexity ≤7 requirements.
- **Comprehensive type handling**: Covers all Go types including generics
  and interfaces.
- **Memory-efficient**: Field-aligned test case structs with factory
  functions.
- **Full test coverage**: 97.2% coverage with 100+ test scenarios.

### Special Cases Handled

- **Zero-capacity slices**: Go optimises `make([]T, 0)` to share memory.
  IsSame treats these as distinct to preserve expected semantics.
- **Invalid reflect.Values**: Returns false rather than panicking when
  comparing invalid reflect.Values.
- **Unsafe pointers**: Full support with appropriate DeepSource
  annotations.
- **Complex numbers**: Proper handling including NaN and Inf edge cases.

### Files Changed

| File | Changes | Description |
|------|---------|-------------|
| `zero.go` | +198 lines | Core `IsSame()` implementation with helpers |
| `testing.go` | +107 lines | Assert/AssertMust functions and documentation |
| `zero_test.go` | +540 lines | Comprehensive test suite with all edge cases |
| `testing_test.go` | +150 lines | Assertion function tests using MockT |
| `compounderror_test.go` | 118 changes | Method chaining and error validation |
| `errgroup_test.go` | 16 changes | Error instance verification |
| `lists_test.go` | 4 changes | Copy independence check |
| `as_test.go` | 12 changes | Slice length validation |
| `README.md` | +33 lines | Same-ness functions and patterns |
| `TESTING.md` | +2 lines | Updated assertion function list |

## Usage Examples

```go
// Value types - same when equal
IsSame(42, 42)           // true
IsSame("hello", "hello") // true

// Reference types - same when same pointer
slice1 := []int{1, 2, 3}
slice2 := slice1         // same reference
slice3 := []int{1, 2, 3} // different reference, same content

IsSame(slice1, slice2)   // true  - same pointer
IsSame(slice1, slice3)   // false - different pointers

// Empty slices - special case
empty1 := make([]int, 0)
empty2 := make([]int, 0)
IsSame(empty1, empty2)   // false - separate make() calls

// Testing assertions
AssertSame(t, slice1, slice2, "slice reference")
AssertNotSame(t, slice1, slice3, "different slices")

// Fatal assertions (call t.FailNow() on failure).
AssertMustSame(t, slice1, slice2, "critical slice reference")
AssertMustNotSame(t, slice1, slice3, "critical different slices")
```

## Verification

- ✅ **All tests pass**: No functionality broken.
- ✅ **Full linting compliance**: 0 issues reported.
- ✅ **High test coverage**: 97.2% coverage maintained.
- ✅ **Documentation complete**: README and TESTING.md updated.
- ✅ **Pattern consistency**: Follows established testing guidelines.

## Why This Matters

The existing `AssertEqual()` uses deep comparison, which doesn't work for
testing same-ness of reference types like maps and slices. These new
functions provide:

1. **Clearer intent**: Explicitly test for reference equality vs value
   equality.
2. **Better error messages**: Consistent formatting and logging.
3. **Modern patterns**: Follows documentation standards for test structure.
4. **Type safety**: Handles all Go types correctly including nil cases.
5. **Safety**: Protection against panics from invalid reflect.Values.

This enhancement strengthens the testing foundation for both internal
library tests and external library users.

___

### **PR Type**

Enhancement, Tests

___

### **Description**

- Add `IsSame()`, `AssertSame()`, and `AssertNotSame()` functions.
- Add `AssertMustSame()` and `AssertMustNotSame()` fatal assertion variants.
- Document cross-compatible test function patterns for `*testing.T` and `MockT`.
- Add comprehensive doRun helper pattern examples and usage guidance.
- Modernize test patterns across 4 test files.
- Replace manual pointer comparisons with new assertions.
- Convert `t.Fatal()` patterns to `if !Assert() { return }`

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["IsSame() function"] --> B["AssertSame() assertion"]
  A --> C["AssertNotSame() assertion"]
  B --> D["AssertMust variants"]
  C --> D
  D --> E["Cross-compatible patterns"]
  E --> F["Test pattern improvements"]
  F --> G["Better error messages"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td><strong>zero.go</strong><dd><code>Add IsSame function with helper functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/core/pull/135/files#diff-670566f7c438f1e9ac53678f74c5a8fbd2eb6d5e4c2a0a010558f203e93bed1a">+127/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>testing.go</strong><dd><code>Add AssertSame and AssertNotSame functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/core/pull/135/files#diff-0d2abf7782555c940e256bd9fa2c45c5522e77102e8d7cad03b094cbf6bef46f">+62/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>as_test.go</strong><dd><code>Replace manual length checks with AssertEqual</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/core/pull/135/files#diff-2bb9510614f1f9ca229f1129aaf952c611523affd3fe5ce0dea3c66c7154ba2d">+6/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>compounderror_test.go</strong><dd><code>Modernize test patterns and error validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/core/pull/135/files#diff-098c7bc3ff8b28639254a2ec917c7f34a810189ab1b121014fcc492d5e94f176">+42/-53</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>errgroup_test.go</strong><dd><code>Replace manual error checks with AssertSame</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/core/pull/135/files#diff-10934bcd7b4b660dfba6fb9312efa594b5ab4568838f302658cc10a0b44f6bc4">+4/-12</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>lists_test.go</strong><dd><code>Replace pointer comparison with AssertNotSame</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/core/pull/135/files#diff-eac72f7cf4789668c68b5fd93e5e2231b34902b18ca037d18ca1fe55cf13dbaa">+1/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>zero_test.go</strong><dd><code>Add comprehensive IsSame test suite</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/core/pull/135/files#diff-de67a8abadf082c8786d9a49109c6392822763101ec655504ae2587e6282173c">+211/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>testing_test.go</strong><dd><code>Add tests for new assertion functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/core/pull/135/files#diff-954b3e0dfe3e1a2179981e401950213ab7c3728acd19c01f6275c0dcb3047eae">+94/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>README.md</strong><dd><code>Document new IsSame and assertion functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/core/pull/135/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5">+11/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>TESTING.md</strong><dd><code>Add new assertion functions to list</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/core/pull/135/files#diff-4da5f2fd7bc2f11672282410e99aad91d422f9f4fb6c723abc7ef29ee5cb6906">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

## Summary by CodeRabbit

* **New Features**
  * Added IsSame (value vs reference identity), a generic S[...] slice helper, AssertSame/AssertNotSame and many AssertMust... fatal variants.

* **Documentation**
  * Updated README and testing guidelines with sameness semantics and cross-compatible Run() patterns.

* **Tests**
  * Extensive IsSame coverage, recursion/edge-case tests; refactored many tests to use the new assertion helpers.

* **Chores**
  * Added "skipcq" to spellcheck dictionary.

### CODE REVIEW

File: compounderror_test.go [hunk 1]
```diff
@@ -375,9 +367,7 @@ func (tc compoundErrorAppendTestCase) Test(t *testing.T) {
 	result := ce.Append(tc.err, tc.note, tc.args...)
 
 	// Test method chaining
-	if result != ce {
-		t.Fatalf("expected same CompoundError instance for chaining")
-	}
+	AssertSame(t, ce, result, "CompoundError instance")
```
[the remark this answers is not in the document]
Comment (amery) [reply compounderror_test.go:370]: redundant because it's the end of the function

**Comment by amery** [comment]
/review
--pr_reviewer.inline_code_comments=true

File: zero.go [hunk 2]
```diff
@@ -169,3 +174,143 @@ func isReflectValueNil(v reflect.Value) bool {
...
+		a := va.Elem().Interface()
+		b := vb.Elem().Interface()
+		ok = IsSame(a, b)
+	case reflect.String, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr,
+		reflect.Float32, reflect.Float64, reflect.Complex64, reflect.Complex128, reflect.Bool:
+		ok = va.Interface() == vb.Interface()
+	}
+	return ok
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply zero.go:316]: Can you imagine a test case that could cause a deep recursion? I couldn't

[the remark this answers is not in the document]
Comment (amery) [reply zero.go:316]: @qodo-merge-tools f2bf707 adds your test cases, they work just fine, confirming what @coderabbitai said

**Comment by amery** [comment]
/review
--pr_reviewer.inline_code_comments=true
-i

File: zero.go [hunk 3]
```diff
@@ -169,3 +174,143 @@ func isReflectValueNil(v reflect.Value) bool {
...
+	case reflect.Interface:
+		// Extract concrete values and compare them recursively
+		a := va.Elem().Interface()
+		b := vb.Elem().Interface()
+		ok = IsSame(a, b)
+	case reflect.String, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr,
+		reflect.Float32, reflect.Float64, reflect.Complex64, reflect.Complex128, reflect.Bool:
+		ok = va.Interface() == vb.Interface()
+	}
```
[the remark this answers is not in the document]
Comment (amery) [reply zero.go:314]: that would test equality, not sameness

[the remark this answers is not in the document]
Comment (amery) [reply zero.go:314]: what do you think it would be more idiomatic for sameness of structs and arrays passed as value? always false or content equality?

**Comment by amery** [comment]
## PR Update: Rebased and Additional Tests Added

This PR has been rebased onto the latest main branch and updated with additional test coverage as requested.

### Changes in this update:

1. **Added unsafe.Pointer test cases** - Comprehensive tests confirming that unsafe pointers are correctly compared by address, including nil handling.

2. **Added complex number test cases** - Tests for both `complex64` and `complex128` types, including edge cases with NaN and Inf values to verify IEEE 754 semantics.

3. **Added default case to isSamePointer** - Improves robustness by explicitly handling unexpected reflect.Kind values.

### Test coverage:
- All new tests pass successfully.
- Coverage remains at 97.2%.
- Tests confirm correct behaviour for all supported types.

Thank you @karasz for the review and suggestion to add these test cases for completeness.

File: zero.go [hunk 4]
```diff
@@ -169,3 +174,151 @@ func isReflectValueNil(v reflect.Value) bool {
...
```
[the remark this answers is not in the document]
Comment (amery) [reply zero.go:322]: give me a test-case that fails. `ok` is `false` unless explicitly made `true`

**Comment by amery** [comment]
> This looks awesome as usual. However just for the sake of completeness don't you think that we should have at least:

@karasz added, they all passed :)

File: zero.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (amery) [reply zero.go:322]: the test failed. thanks

File: zero.go [hunk 5]
```diff
@@ -169,3 +174,151 @@ func isReflectValueNil(v reflect.Value) bool {
...
+		return false
+	}
+
+	if same, ok := isSameTypedNil(va, vb); ok {
+		// typed nil decision
+		return same
+	}
+
+	return isSamePointer(va, vb)
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply zero.go]: I forgot to push

File: zero_test.go [hunk 6]
```diff
@@ -877,3 +881,460 @@ func testZeroEdgeCasesNilFuncPointer(t *testing.T) {
...
+			"functions with different references should not be same"),
+
+		// Pointers - compared by address
+		newIsSameTestCase("same pointer reference", ptr1, ptr2, true,
+			"pointers with same reference should be same"),
+		newIsSameTestCase("pointers to same address", ptr1, ptr3, true,
+			"pointers to same address should be same"),
+	)
+}
+
```
[the remark this answers is not in the document]
Comment (amery) [reply zero_test.go]: i forgot to push

**Comment by karasz** [comment]
> @karasz added, they all passed :)


:) phewww! my mojo seems not that bad as it should.
